utils/s3/client.cc

```diff
@@ -854,6 +858,167 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+    const size_t _part_size;
+    const size_t _total_size;
+    std::optional<future<>> _uploading;
+
+    // avoid stuffing the memory with yet-uploaded buffers in the expense of
+    // having ping-pong between reader (from disk) and writer (to the wire)
+    seastar::pipe<temporary_buffer<char>> _part_body{16};
+
+    future<> upload_part(size_t total_offset, size_t part_size) {
+        // upload a part in a multipart upload, see
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        unsigned part_number = _part_etags.size();
+        _part_etags.emplace_back();
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        assert(total_offset + part_size <= _total_size);
+        if (total_offset + part_size == _total_size) {
+            assert(part_size > 0);
+        } else {
+            assert(part_size >= aws_minimum_part_size);
+        }
+        req._headers["Content-Length"] = to_sstring(part_size);
+        req.query_parameters.emplace("partNumber", to_sstring(part_number + 1));
+        req.query_parameters.emplace("uploadId", _upload_id);
+        s3l.trace("PUT part {} {}~{}/{} bytes (upload id {})",
+                  part_number, total_offset, part_size, _total_size, _upload_id);
+        req.write_body("bin", part_size, [part_size, this] (output_stream<char>&& out_) mutable -> future<> {
+            auto out = std::move(out_);
+            std::exception_ptr ex;
+            try {
+                for (size_t part_offset = 0, written = 0; part_offset < part_size; part_offset += written) {
+                    auto buffer = co_await _part_body.reader.read();
+                    if (!buffer.has_value()) {
+                        // the writer side is closed, close() will abort this
+                        // multipart upload.
+                        break;
+                    }
+                    if (size_t part_remainder = part_size - part_offset; buffer->size() > part_remainder) {
+                        auto front = buffer->share(0, part_remainder);
+                        buffer->trim_front(part_remainder);
+                        // return the remainder of the buffer back to the pipe, so the next
+                        // part can pick it up
+                        _part_body.reader.unread(std::move(*buffer));
+                        co_await out.write(front.get(), front.size());
+                        written = part_remainder;
+                    } else {
+                        written = buffer->size();
+                        co_await out.write(buffer->get(), buffer->size());
+                    }
+                }
+                co_await out.flush();
+            } catch (...) {
+                ex = std::current_exception();
+            }
+            co_await out.close();
+            if (ex) {
+                std::rethrow_exception(std::move(ex));
+            }
+        });
+        auto holder = _bg_flushes.hold();
```

Comment:
I'm not quite sure this is needed. You wait for the whole uploader fiber to resolve in close(), so the bacground gate becomes a no-op. No?